### ISSUE TITLE
Refactor Definition of RPC Commands

### ISSUE THREAD

**Issue by ejmr** [emacs-php/phpactor.el#52]
See the code in the `refactor-rpc-functions` branch for all of the details.  *The code is not finished.*  But while looking at the project it struck me as a potentially useful refactoring.  Please do not merge it because it needs more testing, and discussion, as not everyone may agree it's a worthwhile change.

**Comment by ejmr** [comment]
I should have included a link to the commit in question: bc42e636793e820241939dedbb4ee81163a212ed, to give an idea of my thought process.

**Comment by kermorgant** [comment]
I've never used defmacro before, and have no strong opinion about this change as I'm still a elisp beginner. Just a few questions, mainly out of curiosity 

* How does that affect debugging : are the macro defined functions as easily debuggable ?
* The replaces functions had autoloading enabled. How does that change affect that ? Should that be a concern ?

I'll try to test this branch in the coming days so I'll probably find out the answer to the first question by myself then ;-)

**Comment by ejmr** [comment]
> Are the macro defined functions as easily debuggable?

No.  But that doesn't mean they are difficult to debug either.  Personally the way I debug anything dealing with macros is by using `(macroexpand ...)`, which lets you see exactly what code a macro expands into.  It lets me verify that the macro is creating the code I expect.  However it must be said that if developers run into a problem with such a function, GNU Emacs won't point them towards the macro, because for intents and purposes the macro no longer "exists" when the code is compiled.

> The replaced functions had autoloading enabled.  How does that change affect that?

The `dolist` form that invokes the macro is marked with the `;;;###autoload` directive, so it should apply to every function create by the macro in that `dolist` construct.

But again, this is all just the result of some brainstorming.  Whenever I see a bunch of nearly-identical functions in code, *especially* in a Lisp with powerful macros, it sets off an alarm in my head reminding me that functional programming and macros provide a way to remove that duplication and (almost) always in a way that decreases bugs and improves how easy it is to add new functionality in the future.
